### DISCUSSION TITLE
Application Analytics

### DISCUSSION THREAD

**Discussion by AtjonTV** [kuzzleio/kuzzle#1931] [Ideas]
Hello!

I am currently in the search of a BaaS with features similar to Firebase.

This includes User Management (Login, Perms), Data Storage (at least JSON), App Analytics and Error Tracking.

From what I found after 3 days of searching on Google and Searx, I found a few Promising looking Platforms that all had some issue or another. (From outdated UI, to incomplete features over missing features to having a last commit in 2015)

As I dislike using Google (Except for search), using Firebase itself is out of questions.

Now from what I can see, Kuzzle has 2 out of the 4 features I desire. (User and Data).

But what I am missing is Analytics, from what I can read in the Docs I can make a Plugin that adds a API that I can call from Apps to insert analytical data. The same goes for Error Tracking, though that could be redirected to Sentry. (Note that I also want to go away from Sentry because its huge, slow and doesnt do much yet needs quite some attention)

My point in coming here is to ask if there are any Plans or Ideas at Kuzzle to implement App Analytics for mobile apps (Not for the Kuzzle Server, as that is covered by existing Plugins).

What I mean with Analytics, I think of the following:

* Collecting Data via an API
* Storing data in Time Series
* Ability to Query and Graph Data

The first can be done via a simple API Controller, yet there has to be a format of how the data comes in.

The second is storing the data (which comes with a timestamp) and storing it with the timestamp as primary search index.

Third could be done via Grafana or in the Kuzzle Dashboard. As from what I understand Kuzzle uses Elasticsearch as storage, Grafana could possiby directly get the data out.

I am just here to ask, and bring Ideas. While I do program in Java Script I am not familiar with neither NodeJS nor Elasticsearch.

**Comment by Aschen** [comment]
Hi @AtjonTV !

We do not plan to develop a built-in feature to collect mobile application analytics because needs can be very different between users.

But as you mention, it's very easy to implement an analytics system with Kuzzle without code:
  - create a `analytics` collection with the desired data mappings (deviceId, platformVersion, etc.)
  - add only `document:create` permissions on the collection to logged users 
  - each document is created with [metadata](https://docs.kuzzle.io/core/2/guides/main-concepts/data-storage/#kuzzle-metadata) which contain the creation timestamp
  - Kibana or Grafana can be plugged into Elasticsearch to display dashboards (we also have an article with [freeboard](https://blog.kuzzle.io/how-to-create-a-realtime-iot-dashboard-for-kuzzle))
  
Kuzzle is extensible like any other framework, you can learn how to extend Kuzzle here: https://docs.kuzzle.io/core/2/guides/getting-started/write-application/

**Reply by AtjonTV** [reply]
Thanks!

If it is as easy as you say, I will 100% try Kuzzle hands-on tonight and find my way into it.

The only thing that somewhat irritates me to modify code of Kuzzle, or write plugins, is the fact it uses Elasticsearch as data storage .. specially with PostgreSQL having so many awesome features (like indexing JSON and performing searches in JSON and Fulltext).

(I might be a little bit biased towards using PostgreSQL because all my experience and research just showed me it can do 90% of what any application might need, with only a few edge cases where other systems are more suitable. I run 80% of all my self-hosted apps with PostgreSQL if possible.)

**Reply by AtjonTV** [reply]
Okay .. my local test starts out "great" .. took the docker-compose and tried to just "run" it .. but I cant save anything because everything is write and delete locked in Elasticsearch.

I cant even remove the delete lock because I cant run delete on the lock :disappointed: 

![image](https://user-images.githubusercontent.com/24845056/104222160-2e856180-5442-11eb-945a-d823f03699a0.png)
(Trying to issue a delete on the lock brings the same error message just for the block index thingy)

**Reply by AtjonTV** [reply]
Okay .. soo .. when I manually download the docker-compose that is referenced in the installer script, Elasticsearch does not allow adding to indexes.

But when I clone the kuzzle-plugin-cluster repo and run it from there, it seems to work perfectly fine ..

**Reply by AtjonTV** [reply]
Nvm I eat my own words, I can create a Index and a Collection but I cant add Documents to it
